File: book/petrol-pump/chapter2/section9/CapturePrice.java
package chapter2.section9;

import pump.*;
import sodium.*;

public class CapturePrice implements Pump
{
    public Outputs create(Inputs inputs) {
        return new Outputs()
            .setPriceLCD1(perFuel(inputs.eNozzle1, inputs.price1))
            .setPriceLCD2(perFuel(inputs.eNozzle2, inputs.price2))
            .setPriceLCD3(perFuel(inputs.eNozzle3, inputs.price3));
    }

    private static Behavior<String> perFuel(
                Event<UpDown> eNozzle, Behavior<Double> price) {
        Behavior<Double> capPrice = eNozzle.snapshot(price).hold(0.0);
        Behavior<UpDown> nozzle = eNozzle.hold(UpDown.DOWN);
        return Behavior.lift(
            (u, price_) ->
                u.equals(UpDown.UP) ? Formatters.formatPrice(price_)
                                    : "",
            nozzle, capPrice);
    }
}

File: c#/Book/PetrolPump/PetrolPump/PetrolPump/Chapter2/Section11/CapturePrice.cs
using System;

using PetrolPump.Pump;

using Sodium;

namespace PetrolPump.Chapter2.Section11
{
  public class CapturePrice : IPump
  {
    public Outputs Create(Inputs inputs)
    {
      return new Outputs()
        .SetPriceLcd1(perFuel(inputs.ENozzle1, inputs.Price1))
        .SetPriceLcd2(perFuel(inputs.ENozzle2, inputs.Price2))
        .SetPriceLcd3(perFuel(inputs.ENozzle3, inputs.Price3));
    }

    static Behavior<String> perFuel(
      Event<UpDown> eNozzle,
      Behavior<Double> price)
    {
      Behavior<Double> capPrice = eNozzle.Snapshot(price).Hold(0.0);
      Behavior<UpDown> nozzle = eNozzle.Hold(UpDown.DOWN);
      return Behavior<UpDown>.Lift(
        (u, price_) => u.Equals(UpDown.UP) ? Formatters.FormatPrice(price_) : "",
        nozzle,
        capPrice);
    }
  }
}